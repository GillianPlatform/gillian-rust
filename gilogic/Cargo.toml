[package]
name = "gilogic"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
gilogic_proc = { path = "../gilogic_proc" }
gilogic_proc_dummy = { path = "../gilogic_proc_dummy" }