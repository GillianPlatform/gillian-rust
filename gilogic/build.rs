fn main() {
    println!("cargo::rustc-check-cfg=cfg(gillian)");
}
