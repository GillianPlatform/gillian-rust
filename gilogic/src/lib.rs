#![allow(internal_features)]
#![feature(rustc_attrs)]
#![feature(never_type)]
#![feature(register_tool)]
#![feature(ptr_internals)]
#![feature(allocator_api)]
#![feature(tuple_trait, unboxed_closures)]
#![feature(stmt_expr_attributes, proc_macro_hygiene)]
#![register_tool(gillian)]

mod tys;

pub use tys::RustAssertion;

pub mod macros {
    pub use gilogic_proc::{
        assertion, assertion_test, borrow, close_borrow, extract_lemma, lemma, open_borrow,
        predicate, show_safety, specification,
    };
    pub mod prophecies {
        pub use gilogic_proc::with_freeze_lemma_for_mutref_pcy as with_freeze_lemma_for_mutref;
    }
    pub mod no_prophecies {
        pub use gilogic_proc::with_freeze_lemma_for_mutref_no_pcy as with_freeze_lemma_for_mutref;
    }
}

mod seq;
pub use seq::Seq;

mod ownable;
pub use ownable::Ownable;
pub mod alloc;
pub mod iterated;
pub mod prophecies;

pub mod symex_ctrl {

    #[gillian::builtin]
    #[gillian::no_translate]
    #[rustc_diagnostic_item = "gillian::symex_ctrl::symex_next"]
    #[inline(always)]
    pub fn symex_next() {}
}

#[path = "stubs.rs"]
pub mod __stubs;
