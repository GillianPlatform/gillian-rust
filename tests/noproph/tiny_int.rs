//@check-pass
extern crate gilogic;

use gilogic::{macros::*, Ownable};

struct MyInt(u32);

impl Ownable for MyInt {
    #[predicate]
    fn own(self) {
        assertion!((0 <= self.0) * (self.0 <= 10))
    }
}

impl MyInt {
    #[show_safety]
    fn new(x: u32) -> Option<Self> {
        if x > 10 {
            None
        } else {
            Some(MyInt(x))
        }
    }
}
