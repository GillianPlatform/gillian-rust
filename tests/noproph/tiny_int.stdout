#begin_init_data
{}
#end_init_data

import "i__binop.gil", "i__lang.gil", "i__std_shims.gil", "i__prophecies.gil", "i__ownable.gil", "i__std_shims_no_pcy.gil";

pred "<MyInt as gilogic::Ownable>::own"(+self):
(0i i<=# l-nth(#self, 0i)) * (l-nth(#self, 0i) i<=# 10i) * (self == #self);


proc "MyInt::new"(m_1x) {
  bb0:  m_3 := m_1x;
        g_0 := m_3;
        m_2 := (not (g_0 i<= 10i));
        g_1 := m_2;
        goto [g_1] bb1 bb2;
  bb1:  m_0 := {{ 0i, {{  }} }};
        goto bb3;
  bb2:  m_5 := m_1x;
        g_2 := m_5;
        m_4 := {{ g_2 }};
        g_3 := m_4;
        m_0 := {{ 1i, {{ g_3 }} }};
        goto bb3;
  bb3:  g_4 := m_0;
        ret := g_4;
        goto rlab;
  rlab: return
};

